File: Differential-Equations/tutor/adaptive.py
"""Adaptive engine: ELO ratings, topic unlocking, and problem selection."""

from __future__ import annotations

import random
from datetime import datetime, timezone
from typing import Optional

from .db import ALL_TOPICS, get_overdue_subtopics

# ── ELO constants ─────────────────────────────────────────────────────────────

K_FACTOR = 24.0
ELO_PER_DIFF = 120.0
ELO_D1_MID = 480.0
UNLOCK_ELO = 620.0

# Problem-type difficulty modifiers: MC/TF are easier → lower item ELO
PTYPE_OFFSET: dict[str, float] = {
    "symbolic":       0.0,
    "numeric":      -30.0,
    "true_false":   -50.0,
    "multiple_choice": -70.0,
}

# ── Prerequisite graph ────────────────────────────────────────────────────────

PREREQUISITES: dict[str, list[str]] = {
    "analysis":      [],
    "linalg":        [],
    "multivariable": ["analysis", "linalg"],
    "vectors":       ["multivariable"],
    "odes":          ["analysis", "linalg"],
    "fourier":       ["odes"],
    "pdes":          ["fourier", "vectors"],
    "complex":       ["multivariable"],
}

TOPIC_LABELS = {
    "analysis":      "Real Analysis",
    "linalg":        "Linear Algebra",
    "multivariable": "Multivariable Calculus",
    "vectors":       "Vector Calculus",
    "odes":          "Ordinary Differential Equations",
    "fourier":       "Fourier Analysis",
    "pdes":          "Partial Differential Equations",
    "complex":       "Complex Analysis",
}


# ── ELO maths ─────────────────────────────────────────────────────────────────

def difficulty_elo(difficulty: int, problem_type: str = "symbolic") -> float:
    base = ELO_D1_MID + (difficulty - 1) * ELO_PER_DIFF
    return base + PTYPE_OFFSET.get(problem_type, 0.0)


def expected_score(player_elo: float, item_elo: float) -> float:
    return 1.0 / (1.0 + 10.0 ** ((item_elo - player_elo) / 400.0))


def update_elo(player_elo: float, difficulty: int, actual: float,
               problem_type: str = "symbolic") -> float:
    """
    actual: 1.0=correct, 0.5=correct-with-hint, 0.0=wrong.
    problem_type adjusts the item ELO so easier formats yield less gain.
    """
    item_elo = difficulty_elo(difficulty, problem_type)
    exp = expected_score(player_elo, item_elo)
    return max(100.0, min(1200.0, player_elo + K_FACTOR * (float(actual) - exp)))


def elo_to_difficulty(elo: float) -> int:
    d = int((elo - ELO_D1_MID) / ELO_PER_DIFF) + 1
    return max(1, min(5, d))


def elo_to_pct(elo: float) -> float:
    return max(0.0, min(100.0, (elo - 100.0) / (1200.0 - 100.0) * 100.0))


# ── Topic availability ────────────────────────────────────────────────────────

def unlocked_topics(mastery: dict[str, float]) -> list[str]:
    return [
        t for t in ALL_TOPICS
        if all(mastery.get(p, 0.0) >= UNLOCK_ELO
               for p in PREREQUISITES.get(t, []))
    ]


# ── Problem selection ─────────────────────────────────────────────────────────

def select_topic(mastery: dict[str, float], last_topic: Optional[str],
                 stats: Optional[list[dict]] = None,
                 profile: str = "default") -> str:
    available = unlocked_topics(mastery)
    if len(available) == 1:
        return available[0]

    candidates = [t for t in available if t != last_topic] or available

    now = datetime.now(timezone.utc)
    weights = []
    for t in candidates:
        elo = mastery.get(t, 450.0)
        pct = elo_to_pct(elo) / 100.0

        recency = 1.0
        if stats:
            row = next((r for r in stats if r["topic"] == t), None)
            if row and row.get("last_seen"):
                try:
                    ls = datetime.fromisoformat(row["last_seen"])
                    if ls.tzinfo is None:
                        ls = ls.replace(tzinfo=timezone.utc)
                    hours = (now - ls).total_seconds() / 3600.0
                    recency = 1.0 + min(hours / 24.0, 2.0)
                except ValueError:
                    pass

        # Boost topics with overdue SRS subtopics
        overdue = get_overdue_subtopics(profile, t)
        srs_boost = 1.0 + 0.5 * min(len(overdue), 4)

        weight = recency * srs_boost * (1.0 - 0.7 * pct + 0.05)
        weights.append(weight)

    return random.choices(candidates, weights=weights, k=1)[0]


def select_difficulty(elo: float) -> int:
    """70% at natural level, 20% one below (review), 10% one above (challenge)."""
    base = elo_to_difficulty(elo)
    roll = random.random()
    if roll < 0.70:
        return base
    elif roll < 0.90:
        return max(1, base - 1)
    else:
        return min(5, base + 1)

File: Differential-Equations/tutor-rs/src/adaptive.rs
use rand::Rng;
use crate::db::MasteryRow;
use crate::problems::types::Topic;

// ── ELO constants ─────────────────────────────────────────────────────────────

pub const K: f64 = 24.0;
pub const D1_MID: f64 = 480.0;
pub const ELO_PER_DIFF: f64 = 120.0;
pub const UNLOCK_ELO: f64 = 620.0;

// ── ELO math ──────────────────────────────────────────────────────────────────

pub fn problem_elo(difficulty: u8, type_offset: f64) -> f64 {
    D1_MID + (difficulty as f64 - 1.0) * ELO_PER_DIFF + type_offset
}

pub fn expected_score(player_elo: f64, item_elo: f64) -> f64 {
    1.0 / (1.0 + 10f64.powf((item_elo - player_elo) / 400.0))
}

pub fn update_elo(player_elo: f64, difficulty: u8, actual: f64, type_offset: f64) -> f64 {
    let item_elo = problem_elo(difficulty, type_offset);
    let exp = expected_score(player_elo, item_elo);
    (player_elo + K * (actual - exp)).clamp(100.0, 1200.0)
}

// ── Difficulty / mastery helpers ──────────────────────────────────────────────

pub fn elo_to_difficulty(elo: f64) -> u8 {
    let d = ((elo - D1_MID) / ELO_PER_DIFF) as i32 + 1;
    d.clamp(1, 5) as u8
}

pub fn elo_to_pct(elo: f64) -> f64 {
    ((elo - 100.0) / (1200.0 - 100.0) * 100.0).clamp(0.0, 100.0)
}

/// 70% at natural difficulty, 20% one below, 10% one above.
pub fn select_difficulty(elo: f64, rng: &mut impl Rng) -> u8 {
    let base = elo_to_difficulty(elo);
    let roll: f64 = rng.gen();
    if roll < 0.70 {
        base
    } else if roll < 0.90 {
        base.saturating_sub(1).max(1)
    } else {
        (base + 1).min(5)
    }
}

// ── Topic availability ────────────────────────────────────────────────────────

pub fn unlocked_topics(mastery: &[MasteryRow]) -> Vec<Topic> {
    let elo_of = |t: &Topic| -> f64 {
        mastery.iter()
            .find(|m| m.topic == t.as_str())
            .map(|m| m.elo)
            .unwrap_or(400.0)
    };

    Topic::all().iter()
        .filter(|t| {
            t.prerequisites().iter().all(|prereq| elo_of(prereq) >= UNLOCK_ELO)
        })
        .cloned()
        .collect()
}

// ── Topic selection ───────────────────────────────────────────────────────────

pub fn select_topic(
    mastery: &[MasteryRow],
    unlocked: &[Topic],
    last_topic: Option<&str>,
    rng: &mut impl Rng,
) -> Topic {
    if unlocked.len() == 1 {
        return unlocked[0].clone();
    }

    let candidates: Vec<&Topic> = unlocked.iter()
        .filter(|t| Some(t.as_str()) != last_topic)
        .collect();
    let candidates = if candidates.is_empty() { unlocked.iter().collect::<Vec<_>>() } else { candidates };

    let now = chrono::Utc::now();
    let weights: Vec<f64> = candidates.iter().map(|t| {
        let row = mastery.iter().find(|m| m.topic == t.as_str());
        let elo = row.map(|r| r.elo).unwrap_or(400.0);
        let pct = elo_to_pct(elo) / 100.0;

        // Recency boost: topics not seen recently get higher weight
        let recency = if let Some(Some(last)) = row.map(|r| r.last_seen) {
            let hours = (now - last).num_seconds() as f64 / 3600.0;
            1.0 + (hours / 24.0).min(2.0)
        } else {
            3.0 // never seen: high priority
        };

        recency * (1.0 - 0.7 * pct + 0.05)
    }).collect();

    let total: f64 = weights.iter().sum();
    let mut r: f64 = rng.gen::<f64>() * total;
    for (i, w) in weights.iter().enumerate() {
        r -= w;
        if r <= 0.0 {
            return candidates[i].clone();
        }
    }
    (*candidates.last().unwrap()).clone()
}
